### DISCUSSION TITLE
API Feedback / Clarification

### DISCUSSION THREAD

**Discussion by lenkan** [WebOfTrust/signify-browser-extension#164] [General]
I find the usage of the API a bit unclear. The [docs](https://github.com/WebOfTrust/signify-browser-extension/blob/2c72ac53eb379c9dabde330132378dced01f86bc/GETTING_STARTED_WEB_CLI.md?plain=1#L59-L63) specify something like this for a react app:

```
import { requestAid } from "polaris-web";

export function Component() {
   return <button onClick={requestAid}>Request AID</button>
}
```

I find it a bit confusing what to do with this. I understand there is some sort of subscription that should be done as well. I am wondering if this can be done as part of the polaris library somehow. Something like this would be simpler to use in my opinion.

```
import { requestAid } from "polaris-web";

export function Component() {
   async function handleRequestAid() {
       const result = await requestAid();
       console.log("App received this AID: ", result)
   }
   
   return <button onClick={requestAid}>Request AID</button>
}
```

Do you have any thoughts?

**Comment by HunnySajid** [comment]
> I find the usage of the API a bit unclear. The [docs](https://github.com/WebOfTrust/signify-browser-extension/blob/2c72ac53eb379c9dabde330132378dced01f86bc/GETTING_STARTED_WEB_CLI.md?plain=1#L59-L63) specify something like this for a react app:
> 
> ```
> import { requestAid } from "polaris-web";
> 
> export function Component() {
>    return <button onClick={requestAid}>Request AID</button>
> }
> ```
> 
> I find it a bit confusing what to do with this. I understand there is some sort of subscription that should be done as well. I am wondering if this can be done as part of the polaris library somehow. Something like this would be simpler to use in my opinion.
> 
> ```
> import { requestAid } from "polaris-web";
> 
> export function Component() {
>    async function handleRequestAid() {
>        const result = await requestAid();
>        console.log("App received this AID: ", result)
>    }
>    
>    return <button onClick={requestAid}>Request AID</button>
> }
> ```
> 
> Do you have any thoughts?

Hi, I agree that it might feel a bit confusing initially. 
Let's try to look into more details. 

There are three main actors:
- web script (which is signify-polaris-web)
- content script (The content script is part of extension that is running in context of web page)
- background script (It is a part of extension and can communicate with content script)

Web script communicates with content script in an event-based fashion.

When we call `requestAid` / `requestCredential` / `requestAidOrCredential` from web page it uses
`window.postMessage` API to broadcast events like `select-identifier`, `select-credential` or `select-aid-or-credential`.
We can not `await requestAid` because the workflow continues in the extension and data is not returned immediately.

## Regarding subscription
Before making any call like `requestAid` / `requestCredential` / `requestAidOrCredential`, `subscribeToSignature` needs to be called as it receives a call back that contains data with headers.

## Communication Flow
web script -> extension-content-script -> extension-service-worker

Please review [signify-polaris-web](https://github.com/WebOfTrust/polaris-web/blob/main/index.js). I am open to discuss if there is a better way to implement this. Let's continue the discussion under this thread. 🧵

**Comment by 2byrds** [comment]
@HunnySajid I had a discussion with @lenkan today about the event pub/sub design. Is there a need for having an id with each request so that you know which response matches your request? Does window.postMessage help with that?
Daniel agrees that examples will be helpful. No need to dig into the details of what signify-ts returns, he's very familiar with that. But improving the documentation with examples and flow charts will be great to continue improving this discussion and the 'api'.

**Comment by HunnySajid** [comment]
@2byrds @lenkan 

I have created a [sequence diagram](https://mermaid.live/view#pako:eNq1V21v2zYQ_isEPyWF5dWxHbsKGiBotyEfugV1hwGDgYKSLg4bmdR4lF03yX_vUZRsyZYDpWgTII54vOfuuTedH3isE-AhR_g_BxXDeykWRiznitFPJoyVscyEsmwNUSCy7FCQ6VQYiQFdOBTGWllQNsDYyMy2KWd5CyaCWckYgrU292C83P9tAgaXl82DkFkRzaywwN6yOVdawZx31IQVPbFUIp2CIR8s04pAloAoFlucMhAEUGMeMswjBxTBJz2TCyVsbuAkFmkaifj-JBFWhNdbwenpRUl6B3EUkFlNXiDpyttNgBXGnLO1tHessnHxrHvG5RftlUzY42P19M5AQpylSGuHdOXvj07S6uJ-0NZSJXrdzzTaDz5OJw_MbjIInc-QQmwDWRi5lZRJzp56JHg151UEXpbQFsQSR6T2kJYX_RwW8Q71kMUPM6mjlliQIhwk42cyETIJtPlFhFrAK14qOZLysmZDdq2-EAYymkKpXrj2y4iBr3P4Sp0p1YIVneAk0mAt-X9pNvOSG5Kw3911fIZMJ6OZ0cvMug709Niu8pg2bMeS3dJNP8726dY7ksQh-xciemY5OpAMFBI38g2lM10biO7flqhf4T2yW7Jehf9IVEtr2yRJbG-gTv3zIvSXF_WL4DuUWBW75oskZH-Aje8AmZvHLquGNCgL_1x7teb1Fv13ernMlYzJJfQVUs1lKiDCW0C9RPC3nY94xEJJ9E-5qtxyw9xh3fhSaBIqr-OdXruI1GxRE-_K1JVHrU9coQpL9dV1WLYaq5Ph_X6_67zq4Dm5exy_UyN54k3kQ-Cq2o_Vh8np2hL6SEar-VnsBaEPIXw2gDo3MXz24T3tWDheGxtJoc_IUDSc91Ybt2M82_jUDNjm2ZwbSLVIAiymAT_t-Gb1WtvBJ4s56IZQUdh5ljiPO2IpWKebkmXSYLmSKKMU3Fz1hvaTuQ81KzMpGjBb9Yuj46627mR5RIuc6_T2zcnvegXPYjVju92sde3Zvix2wPt6277NDKykzpHisd3hkvqa5g3wHm2WZknDjNbvB3c25_YOluSeS2kizL2r1ie6J3KrZxsV89CaHHrcp6Zc1atDWpz_07r-yMMH_pWHg8Hr_uBsMJyOBpPJZDw-n_T4hodno_749Xg6HJ2NBuPhaDoYPfX4twJh2B-fT8_e0O_wfHQ-mb7pcUgklegH_2Wh-M7w9B1EF0CL) to elaborate the workflow. 

Also, created [a video to walkthrough the sequence diagram](https://www.loom.com/share/24113545c40041e3bc0c0a74821671f3?sid=b9a81fe8-dfff-4989-a5a0-9f374efcef35). (Please turn subtitles on if inaudible)

I tried to elaborate on the workflow. Let me know your feedback about it.

Also, We can add unique IDs with requests to identify. 
In our example app, we did not find/consider a use case where unique IDs can be required but it would be a good addition.

**Reply by 2byrds** [reply]
That's great @HunnySajid. I believe we need to update the architecture diagram to have the flow go through the content script first before it reaches the background (service worker) script. Do you agree?
https://github.com/WebOfTrust/signify-browser-extension/blob/main/arch_diagram.md

**Reply by rodolfomiranda** [reply]
sequence diagram look correct. I think you can publish to the documentation. We can add the calls to keria in a next version.

**Comment by lenkan** [comment]
I think the new additions is addressing my concerns for this specific discussion. Looks great!
